inginious/frontend/plugins/multilang/problems/custom_test_manager.py

```diff
@@ -0,0 +1,132 @@
+from datetime import datetime
+from bson.objectid import ObjectId
+from pymongo.collection import ReturnDocument
+
+
+class CustomTestManager(object):
+    """ Runs custom test job asynchronously """
+
+    def __init__(self, client, user_manager, database):
+        self._client = client
+        self._user_manager = user_manager
+        self._database = database
+
+    def _job_done_callback(self, custom_test_id, result, stdout, stderr):
+        custom_test = self.get_custom_test(custom_test_id, False)
+
+        if not custom_test:
+            return
+
+        data = {
+            "status": ("done" if result[0] == "success" or result[0] == "failed" else "error"),
+            "result": result[0],
+            "text": result[1],
+            "stdout": stdout,
+            "stderr": stderr
+        }
+
+        unset_obj = {
+            "job_id": "",
+        }
+
+        self._database.custom_tests.find_one_and_update(
+            {"_id": custom_test["_id"]},
+            {"$set": data, "$unset": unset_obj},
+            return_document=ReturnDocument.AFTER
+        )
+
+    def new_job(self, task, input_data, launcher_name="Unknown"):
+        """
+            Runs a new custom test job asynchronously.
+            The general idea works like the SubmissionManager, as it creates a new document in DB to keep track of
+            the running custom tests, When the job is done, it calls a callback to update the document in DB to tell
+            the frontend the job has finished plus the given results.
+            Returns the custom test id.
+        """
+
+        if not self._user_manager.session_logged_in():
+            raise Exception(_("An user must be logged in to run custom tests"))
+
+        username = self._user_manager.session_username()
+
+        # Prevent student from sending several tests together
+        current_custom_test = self._database.custom_tests.find_one({
+            "courseid": task.get_course_id(),
+            "taskid": task.get_id(),
+            "username": username})
+
+        if current_custom_test is not None:
+            # This is in case there is a job in DB that was not previously deleted but already finished.
+            # Also, in case the current custom test is still in waiting status but the job is there for more than 3
+            # minutes because the job probably already finished but the callback was not successfully run.
+            if self.is_done(str(current_custom_test["_id"])) or (
+                    self.is_waiting(str(current_custom_test["_id"])) and self.is_old(current_custom_test)):
+                self.delete_custom_test(str(current_custom_test["_id"]))
```

Comment:
Maybe the parameters on these functions can be written to a variable, since they are the same.